### DISCUSSION TITLE
Question about dependency on "which"

### DISCUSSION THREAD

**Discussion by Lysxia** [srid/neuron#650] [General]
I'm trying to build neuron using plain cabal, instead of via Nix as documented, so I could be missing something. There are a few hacks needed on the way to reproduce the following problem, but basically, if I run `cabal install neuron`, I get the error that `neuron-search`  could not be found while compiling `neuron/src/app/Neuron/CLI/Search.hs`.

In `neuron/src/app/Neuron/CLI/Search.hs`, the function `staticWhich` from the package "which" is used to find the path to the bash script `neuron-search`.

https://github.com/srid/neuron/blob/7d57e45aa0135501975131bf760b04fa9f6bad08/neuron/src/app/Neuron/CLI/Search.hs#L20

I could make it compile using the `Paths_neuron` module instead, but I'm not sure whether that would work for Nix. Does this look like an acceptable fix?

```haskell
neuronSearchScript = $(TH.lift (unsafePerformIO (getDataFileName "src-bash/neuron-search")))
```

https://github.com/Lysxia/neuron/commit/51e5eb075c122b0776864527be34344bb09f943d#diff-156fa4df5d0d36eccfe427a54b100371R22

(Note: I'm actually using tag 0.6.0.0 instead of master, to avoid rib-core, but the relevant line of code hasn't changed so the issue still seems current.)

**Comment by srid** [comment]
May I know the reason for trying to build it with plain cabal, outside of Nix?

Generally speaking, this is best done in the `which` library. There is a PR open for that, but I don't know the status of it: https://github.com/obsidiansystems/which/pull/3

**Comment by Lysxia** [comment]
I don't want to install and learn to use yet another package manager. The required changes to build with vanilla cabal are quite minor; I can volunteer to patch and maintain anything that has to do with this, but I would also understand if you just don't want the extra overhead; this issue is the only one I'm not sure how to fix without breaking stuff.

The linked patch to `which` doesn't seem it would actually work if the required executable is a data file of the current package, because it depends on that executable being on PATH before the package is built.

**Comment by srid** [comment]
Since this is the only place where `staticWhich` is used (it used to be used in several places), then I suppose we can just get rid of it entirely, and possibly even embed the `neuron-search` script in haskell (via `Text.RawString.QQ`).

I don't particularly like seeing unsafePerformIO here (doesn't seem like a conventional solution; is that what all these TH libraries do?).

In any case, whatever change we make should continue to work in nix-build (`./result/bin/neuron -d ./guide search` should not fail), and should not break the compilation in the `bin/run` and `bin/test` scripts (which use ghcid).

**Comment by srid** [comment]
Plus, it should not revert the nix install size optimizations done in https://github.com/srid/neuron/issues/193#issuecomment-646342720 (do a nix install on a fresh Ubuntu VM, and check the size of /nix). I vaguely remember seeing an issue with cabal's Paths_neuron adding unnecessary dependencies.

**Comment by Lysxia** [comment]
> I don't particularly like seeing unsafePerformIO here (doesn't seem like a conventional solution; is that what all these TH libraries do?).

The conventional solution is to use [`runIO`](https://hackage.haskell.org/package/template-haskell-2.16.0.0/docs/Language-Haskell-TH-Syntax.html#v:runIO), I didn't remember the name off-hand.

**Comment by Lysxia** [comment]
Awesome!

**Comment by srid** [comment]
Had to revert that. If you're interested, please see #432

Enough dealing with non-nix requests for me today.

**Comment by srid** [comment]
Closing per https://neuron.zettel.page/next ... to further discuss this idea, in the context of Emanote, [post here](https://github.com/srid/emanote/issues/new/choose)
